extensions/antora/xref-hash-validator/xref-hash-validator.js

```diff
@@ -0,0 +1,142 @@
+const { parse: parseHTML } = require('node-html-parser')
+const { posix: path, resolve } = require('path')
+const url = require('url');
+const http = require("http");
+
+module.exports.register = function ({ config }) {
+
+    let linkTargets = {}
+    
+    const logger = this.getLogger('xref-hash-validator')
+
+    // set a default log level for the log message when an xref is not valid
+    // this an be overridden when the extension is specified in the playbook.yml file
+    const {logLevel = 'warn'} = config
+    
+    this
+    .on('contentClassified', ({ }) => {
+
+        // add a file.xrefChecker property to each file in the contentCatalog
+        // we can add this now, before the contentCatalog object is locked
+        // we can modify its values after the documentsConverted event
+        const { contentCatalog } = this.getVariables()
+        const files = contentCatalog.getFiles().filter((f) => f.src && f.src.mediaType === 'text/asciidoc')
+        files.forEach( (file) => {
+            file.xrefChecker = { 'linkTargets': [], 'internalLinks': []}
+        })
+    })
+    .on('documentsConverted', ({ }) => {
+
+        // after the asciidoc has been converted to HTML we can parse the HTML for link targets
+        // and internal links that target sections of other pages
+        // for each file, we can add this information to file.xrefChecker, which is now part of the contentCatalog
+        const { playbook, contentCatalog } = this.getVariables()
+        const files = contentCatalog.getFiles()
+        files.forEach( (file) => {
+            if (!file.out || !file.asciidoc) return
+
+            // parse the HTML - note that the HTML includes the content of the page only - excluding the document header
+            // so we can't verify any links in the toolbar, nav, footer etc at this point
+            const elements = parseHTML(file.contents.toString()).getElementsByTagName('*')
+            // any element in the HTML that has an id attribute is a valid link target
+            file.xrefChecker.linkTargets = elements.flatMap((e) => {
+                const id = e.getAttribute('id')
+                if (id) {
+                    (linkTargets[id] = linkTargets[id] || []).push({"out": file.pub.url, "src": file.src.path, "component": file.src.component, "version": file.src.version, "branch": file.src.origin.branch});
+                    return [id]
```

Comment:
```suggestion
                    return id
```
These _should_ be equivalent, as `flatMap` flattens one level. 
(I've not tested the change on this PoC, but on a simpler example) 
If that's so, I'd recommend this for simplicity and clarity, as no lists are involved here other than the final mapped output.